Gem/Code/Source/Components/ScriptableDecalComponent.cpp

```diff
@@ -0,0 +1,213 @@
+        if (m_decalFeatureProcessor)
+        {
+            AZ::RPI::Scene* scene = m_decalFeatureProcessor->GetParentScene();
+            DecalRequestBus::Handler::BusConnect(scene->GetId());
+            AZ::TickBus::Handler::BusConnect();
+        }
+    }
+
+    void ScriptableDecalComponent::Deactivate()
+    {
+        AZ::TickBus::Handler::BusDisconnect();
+        DecalRequestBus::Handler::BusDisconnect();
+
+        auto removeDecalsFromList = [&](AZStd::vector<DecalInstance>& container)
+        {
+            for (const DecalInstance& instance : container)
+            {
+                m_decalFeatureProcessor->ReleaseDecal(instance.m_handle);
+            }
+            container.clear();
+        };
+
+        removeDecalsFromList(m_fadingInDecals);
+        removeDecalsFromList(m_fadingOutDecals);
+        removeDecalsFromList(m_decalHeap);
+
+        m_decalFeatureProcessor = nullptr;
+    }
+
+    void ScriptableDecalComponent::SpawnDecal(const AZ::Transform& worldTm, AZ::Data::AssetId materialAssetId, const SpawnDecalConfig& config)
+    {
+        DecalHandle handle = m_decalFeatureProcessor->AcquireDecal();
+
+        AZ::Vector3 scale = AZ::Vector3(config.m_scale, config.m_scale, config.m_scale * config.m_thickness);
+
+        m_decalFeatureProcessor->SetDecalTransform(handle, worldTm, scale);
+        m_decalFeatureProcessor->SetDecalMaterial(handle, materialAssetId);
+        m_decalFeatureProcessor->SetDecalOpacity(handle, config.m_opacity);
+        m_decalFeatureProcessor->SetDecalAttenuationAngle(handle, config.m_attenutationAngle);
+        m_decalFeatureProcessor->SetDecalSortKey(handle, config.m_sortKey);
+
+        uint32_t currentTimeMs = static_cast<uint32_t>(AZ::Interface<AZ::ITime>::Get()->GetElapsedTimeMs());
+
+        if (config.m_fadeInTime > 0.0f)
+        {
+            m_fadingInDecals.push_back({ config, handle, currentTimeMs });
+        }
+        else if (config.m_lifeTime > 0.0f)
+        {
+            uint32_t lifetimeMs = static_cast<uint32_t>(config.m_lifeTime * 1000.0f);
+            uint32_t despawnTimeMs = currentTimeMs + lifetimeMs;
+
+            // Store decals in a min heap sorted by when they need to start animating the fade out.
+            // This makes it very cheap to check each frame if any decals need to change to the fade-out state.
+            m_decalHeap.push_back({ config, handle, despawnTimeMs });
+            AZStd::push_heap(m_decalHeap.begin(), m_decalHeap.end(), HeapCompare);
+        }
+        else
+        {
+            m_fadingOutDecals.push_back({ config, handle, currentTimeMs });
+        }
+    }
+
+    void ScriptableDecalComponent::OnTick([[maybe_unused]] float deltaTime, [[maybe_unused]] AZ::ScriptTimePoint time)
+    {
+        uint32_t currentTimeMs = static_cast<uint32_t>(AZ::Interface<AZ::ITime>::Get()->GetElapsedTimeMs());
```

Comment:
FWIW, AZ::TimeMs is simply an enum class wrapping an int64_t with helpers defined for the various mathematical overloads one would expect of an integral class. It has a wrap time of 290 million years or something like that.. It will always pass by register, so no additional overhead using them as method parameters, it should serialize and bind everywhere you would expect an integral type to, including as a type for AZ::CVars, as a network property, or exposed to script.